rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,623 @@
+   `my_div` Kani reports an error.
+   
+   Unlike a unit-test we can however elide any checks of the output and
+   post-call state. Instead Kani uses the conditions we specified in the
+   contract as checks. Preconditions (`requires`) are inserted as `kani::assume`
+   *before* the call to `my_div`, limiting the generic inputs to those the
+   function is actually defined for. Postconditions (`ensures`) are inserted as
+   `kani::assert` checks *after* the call to `my_div`.
+
+   The expanded version of our harness and function is equivalent to the following:
+
+   ```rs
+   #[kani::proof]
+   fn my_div_harness() {
+     let dividend = kani::any();
+     let divisor = kani::any();
+     kani::assume(divisor != 0); // requires
+     let result = my_div(dividend, divisor);
+     kani::assert(result <= dividend); // ensures
+   }
+   ```
+
+   This expanded harness is then verified like any other harness but also gives
+   the green light for the next step: verified stubbing.
+
+3. In the last phase the **verified** contract is ready for us to use to
+   **stub** other harnesses.
+
+   Unlike in regular stubbing, there has to be at least one associated
+   `proof_for_contract` harness for each function to stub *and* it requires all
+   such harnesses to pass verification before attempting verification of any
+   harnesses that use it as a stub.
+
+   A possible harness that uses our `my_div` contract could be the following:
+
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let some_idx = my_div(v.len() - 1, 3);
+     v[some_idx];
+   }
+   ```
+
+   To use the contract as a stub Kani must first ensure the calling context is
+   safe. It inserts a `kani::assert` for the preconditions (`requires`) before
+   the call; then it replaces the result of `my_div` with a non-deterministic
+   value. Finally the non-deterministic result is constrained by a
+   `kani::assume` of the postconditions (`ensures`).
+   
+   Mutable memory is similarly made non-deterministic, discussed later in
+   [havociing](#memory-predicates-and-havocing).
+
+   Stubbing `my_div` expands it similar to this:
+  
+   ```rs
+   fn my_div_stub(dividend: u32, divisor: u32) -> u32 {
+     kani::assert(divisor != 0); // pre-condition
+     kani::any_where(|result| { /* post-condition */ result <= dividend })
+   }
+   ```
+
+   Notice that this performs no actual computation for `my_div` (other than the
+   conditions) which allows us to avoid something potentially costly.
+
+Also notice that Kani was able to express both contract checking and stubbing
+with existing capabilities; the important feature is the enforcement. The
+checking is, by construction, performed **against the same condition** that is
+later used as stub, which ensures soundness (see discussion on lingering threats
+to soundness in the [future](#future-possibilities) section) and guarding against
+stubs diverging from their checks.
+
```

Comment:
This section describes a lot of low-level details (e.g., the expansions) which in my opinion are more appropriate to discuss in the design section. However, what I'm missing here is the justification to have new APIs and even contract-specific harnesses when there's apparently no need to. Or did I miss something?

Reply:
I think mentioning the expansion was @feliperodri's wish if I recall correctly and I think you have a differing conception of what this section is supposed to contain. I was told this is really addressing the user who cares about how to use the contracts and maybe understanding a bit of how they work. They wouldn't care so much as to the justification for extra harnesses.

I'm not sure what you mean by "new APIs". What specifically are you referring to?

The contract specific harnesses exist because we need *some* way of saying which harness tests which contract and we chose this one. There is a section in "Rationale and Alternatives" discussing the justification which I have yet to fill in. Sorry

Reply:
> I think you have a differing conception of what this section is supposed to contain.

Fair enough.

> What specifically are you referring to?

I'm guessing I was referring to the `proof_for_contract` and `stub_verified` attributes. But let me continue this discussion on the other thread.